### PR TITLE
5182: [TEST] Ha-Flow stats

### PR DESCRIPTION
**Pull request by pkazlenka**
Tests #5182

* Added tests for Ha-Flow stats
* Refactored support for Legacy TSDB stats (from OpenTSDB)
* Added support for VictoriaDB stats

### CODE REVIEW

File: src-java/testing/test-library/src/main/java/org/openkilda/testing/service/tsdb/model/StatsMetric.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2018 Telstra Open Source
+/* Copyright 2223 Telstra Open Source
```
Comment (dmitrii-beliakov) [inline src-java/testing/test-library/src/main/java/org/openkilda/testing/service/tsdb/model/StatsMetric.java:1]: how's life there? OpenKilda is still in use as I see

**Review by dmitrii-beliakov — APPROVED** [review]

File: src-java/testing/test-library/src/main/java/org/openkilda/testing/service/tsdb/model/StatsMetric.java (hunk 1 above)
Comment (pkazlenka) [reply src-java/testing/test-library/src/main/java/org/openkilda/testing/service/tsdb/model/StatsMetric.java:1]: [Timeline](https://www.glow-internet.com/infographics/the-sci-fi-timeline/) is correct enough. After adding AI and moving to cloud, kilda evolved to what you, ancient people, called 'Matrix'. Digital copy of the so-named 'Kilda Architect' still tries to stabilize build, and agents are working (without success) on reducing amount of flaky tests :(

File: src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/model/stats/IslStats.groovy [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        IslStats.statsHelper = statsHelper
+    }
+
+    static IslStats of(Isl isl){
+        return new IslStats(isl)
+    }
+
+    IslStats(Isl isl) {
+        this.metricPrefix = statsHelper.getMetricPrefix()
+//        this.stats = statsHelper.getTsdb().queryDataPointsForLastFiveMinutes(
```
[the remark this answers is not in the document]
Comment (pkazlenka) [reply src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/model/stats/IslStats.groovy]: Done

File: src-java/testing/test-library/src/main/java/org/openkilda/testing/service/tsdb/LegacyTsdbQueryServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    String metricPrefix;
+
+    @Override
+    public List<StatsResult> queryDataPointsForLastFiveMinutes(List<StatsMetric> metrics, String idTag, String id) {
+        return queryDataPointsForLastMinutes(metrics, idTag, id, 5);
+    }
+
+    @Override
+    public List<StatsResult> queryDataPointsForLastMinutes(List<StatsMetric> metrics,
+                                                           String idTag,
```
Comment (yuliiamir) [inline src-java/testing/test-library/src/main/java/org/openkilda/testing/service/tsdb/LegacyTsdbQueryServiceImpl.java:58]: WDYT to add enum StatisticsIdTags for the idTag param ('switchid ', 'ha_flow_id', 'y_flow_id', 'flow_id', etc.)?

Comment (pkazlenka) [reply src-java/testing/test-library/src/main/java/org/openkilda/testing/service/tsdb/LegacyTsdbQueryServiceImpl.java:58]: This is a good idea. I didn't implement it because new (VictoriaDB) API doesn't need it to be passed as separate parameter. This parameter (as well as dozen of other tricks) are added only to unify query interface for VictoriaDB and OpenTSDB. When OpenTSDB is removed, interface would be much simpler. The problem of passing strings as parameters would pass away itself.

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/yflows/YFlowPathSwapSpec.groovy [hunk 4]
```diff
@@ -117,14 +123,26 @@ class YFlowPathSwapSpec extends HealthCheckSpecification {
...
-        statsHelper.verifyYFlowWritesStats(yFlow, beforeTraffic, true)
+        def subflow = yFlow.getSubFlows().shuffled().first()
+        def subflowId = subflow.getFlowId()
+        def dstSwitchId = subflow.getEndpoint().getSwitchId()
+        def flowInfo = database.getFlow(subflowId)
+        def mainForwardCookie = flowInfo.forwardPath.cookie.value
+        def mainReverseCookie = flowInfo.reversePath.cookie.value
+        Wrappers.wait(STATS_LOGGING_TIMEOUT) {
+            def stats = flowStats.of(subflowId)
+            stats.get(FLOW_RAW_BYTES, subflow.getEndpoint().getSwitchId(), mainForwardCookie).hasNonZeroValues()
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/yflows/YFlowPathSwapSpec.groovy:142]: can be replaced by 'dstSwitchId' OR 'dstSwitchId' can be removed

File: src-java/testing/test-library/src/main/java/org/openkilda/testing/service/tsdb/model/LegacyStatsResult.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
+/* Copyright 2223 Telstra Open Source
```
Comment (yuliiamir) [inline src-java/testing/test-library/src/main/java/org/openkilda/testing/service/tsdb/model/LegacyStatsResult.java:1]: typo: 2023

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/stats/HaFlowStatSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    @Shared
+    HaFlowStats haFlowStats
+    @Shared
+    SwitchTriplet switchTriplet
+
+    def setupSpec() {
+        switchTriplet = topologyHelper.getSwitchTriplets(false, false).find {
+            it.ep1 != it.ep2 && it.ep1 != it.shared && it.ep2 != it.shared &&
+                    [it.shared, it.ep1, it.ep2].every { it.traffGens }
+                    && it.ep2.getTraffGens().size() > 1 // needed for update flow test
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/stats/HaFlowStatSpec.groovy:47]: Discussed offline and decided to apply the same approach as for the SwitchPairs. The appropriate task will be created for that and be a part of another PR.
